---
title: Comparison of Output from 2 Amplicon Pipelines- Here I am processing the data
  from the QIIME2 pipeline
author: "Liz Suter"
date: "6/26/2020"
---

Using data from Green et al.:
Green TJ, Siboni N, King WL, Labbate M, Seymour JR, Raftos D. 2019. Simulated Marine Heat Wave Alters Abundance and Structure of Vibrio Populations Associated with the Pacific Oyster Resulting in a Mass Mortality Event. Microb Ecol. 77(3):736â€“747. doi:10.1007/s00248-018-1242-9.

I followed the BVCN lesson 3a and lesson 3b pipelines.
Lesson 3a uses QIIME2, calling DADA2 for ASV assignment, and using a Naive Bayes classifier trained on Silva 132 for taxonomy assignment (data are in `qiime2_export`)
Lesson 3b uses DADA2 directly in R (I used the Cyverse app I created), and IDTAXA (from Decipher) trained in Silva 138 for taxonomy assignment (data are in 'dada2_export')

Modifications:

* There were no blanks so did not remove contamination

### Load packages

```{r}
library(tidyverse)
library(phyloseq)
library(readr)
library(seqinr)
library(decontam)
library(ape)
library(vegan)
library(RColorBrewer)

#install.packages("remotes")
#remotes::install_github("microbiome/microbiome")

library(microbiome)

# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# 
# BiocManager::install("DESeq2")

library("DESeq2")
```

### Import the metadata associated with the samples
These metadata can be used with both qiime2 and dada2 results.

For this Bioproject, all info from Biosample file is also in SraRunTable, so I will only import SraRunTable (it is also formatted better)

I manually modified the SraRunTable to split the sample names into numbers (since temperature was built into the sample name) and put some info about samples from the biosample_result.txt summary (eg. antibiotic treatment). I also added in a column for "mortality" because the paper indicates that oysters in the control samples at 25C faced higher mortality (~50% on day 4 when samples were collected; Fig. 1) and this was not seen in any other samples.

Therefore, I am importing SraRunTableMod.txt

```{r}
SraRunTable <- read_delim("SraRunTableMod.csv", delim = ",")
```


### Import the results from QIIME2

```{r}
# Import Count table. Skip first row of tsv file, which is just some text
count_table <- read_tsv(file="qiime2_export/table/table.tsv", skip = 1)
# And specify that the first column of data are rownames
count_table <- column_to_rownames(count_table, var = colnames(count_table)[1])


# Import taxonomy of ASVs
taxonomy <- read_tsv(file="qiime2_export/taxonomy/taxonomy.tsv")

# Import tree file 
tree = read_tree("qiime2_export/exported-tree/tree.nwk")

# Import fasta
fasta <- read.fasta(file = "qiime2_export/rep-seqs.fasta/dna-sequences.fasta")
```

### Check sequencing depth with rarefaction curves

```{r}
# Use rarecurve, from the Vegan package. Rarcurve expects the dataset as a dataframe so we need to use as.data.frame again:
count_table_df <- as.data.frame(count_table)

# Plot the rarefaction curves, color-coding by the colors listed in sample_info_tab, which indicate sample type, and transforming using t() again
rarecurve(t(count_table_df), step=100, cex=0.5, ylab="ASVs", label=T)

```

Generally these seem to be sequenced to completion

### Remove singletons

```{r}
count_table_no_singletons <- filter(count_table,rowSums(count_table)>1)
# retains 5638 ASVs (out of 5647)
```

### Modify taxonomy table

The taxonomy in the taxonomy table is retained in one column and the different levels are separated by underscore, eg:

```{r}
head(taxonomy)
```

Move each taxonomic level to its own column by removing the ";D_#_" using regular expressions (see this [cheatsheet](https://evoldyn.gitlab.io/evomics-2018/ref-sheets/R_strings.pdf) about regexp)

```{r}
taxonomy_mod <-  taxonomy %>%
  mutate(taxonomy=str_replace_all(string=Taxon, pattern="D_\\d*\\__", replacement="")) %>%
  mutate(taxonomy=str_replace_all(string=taxonomy, pattern=";$", replacement="")) %>%
  separate(taxonomy, into=c("Domain", "Phylum", "Class", "Order", "Family", "Genus","Species"), sep=";") %>%
  select (-Taxon, -Confidence) %>%
  column_to_rownames(var = 'Feature ID') 

head(taxonomy_mod)

```


### Put all into phyloseq object

Here we will do ordinations using the phyloseq package, which first requires making phyloseq objects out of each of our input data tables (in the last tutorial, I imported the tree using phyloseq so it is already a phyloseq object)

```{r}
ASV	=	otu_table(data.frame(count_table_no_singletons), taxa_are_rows =  TRUE)
TAX	=	tax_table(as.matrix(taxonomy_mod))
META	=	sample_data(data.frame(SraRunTable, row.names = SraRunTable$Run))
```

First check that the inputs are in compatible formats by checking for ASV names with the phyloseq function, taxa_names
```{r}
head(taxa_names(TAX))
head(taxa_names(ASV))
head(taxa_names(tree))
```


And check sample names were also detected
```{r}
head(sample_names(ASV))
head(sample_names(META))
```



Make one phyloseq object, which contains all 4 objects:
```{r}
ps <- phyloseq(ASV,	TAX,	META, tree)
```

Check some features of the phyloseq object

```{r}
rank_names(ps)

unique(tax_table(ps)[, "Domain"])

table(tax_table(ps)[, "Domain"], exclude = NULL)

```


Filter out those ambigious Domain annotations
```{r}
ps <- subset_taxa(ps, !is.na(Domain) & !Domain %in% c("Unassigned", "Eukaryota"))

table(tax_table(ps)[, "Domain"], exclude = NULL)
```

Check out the phyla names
```{r}
table(tax_table(ps)[, "Phylum"], exclude = NULL)
```


Filter out any with "NA" as phylum. For now, I am leaving in phyla with abundance of 1 ("Fibrobacteres", "PAUC34f", "Elusimicrobia", "Synergistetes", "Entotheonellaeota", "Lentisphaerae", "Margulisbacteria", "Omnitrophicaeota", "Marinimicrobia (SAR406 clade)")

```{r}
ps <- subset_taxa(ps, !is.na(Phylum) & !Phylum %in% c(""))

table(tax_table(ps)[, "Phylum"], exclude = NULL)
```

Re-root tree (from experience, I have had to do this because you may have removed the root of your tree when pruning). 
(I found this handy function from [here](https://john-quensen.com/r/unifrac-and-tree-roots/) which picks the longest branch to root from). There is also a compatibilty issue between the type of tree calculated by QIIME and the one expected by phyloseq, so we have to change that format as well (see [here](https://github.com/joey711/phyloseq/issues/936) for discussion)
```{r}
# first define function from link above to find furthest outgroup
pick_new_outgroup <- function(tree.unrooted){
require("magrittr")
require("data.table")
require("ape") # ape::Ntip
# tablify parts of tree that we need.
treeDT <- 
     cbind(
         data.table(tree.unrooted$edge),
         data.table(length = tree.unrooted$edge.length)
     )[1:Ntip(tree.unrooted)] %>% 
 cbind(data.table(id = tree.unrooted$tip.label))
 # Take the longest terminal branch as outgroup
 new.outgroup <- treeDT[which.max(length)]$id
 return(new.outgroup) }

# then run on my phyloseq tree
my.tree <- phy_tree(ps)
out.group <- pick_new_outgroup(my.tree)
out.group

# Then use this outgroup to root the tree
new.tree1 <- ape::root(my.tree, outgroup=out.group, resolve.root=TRUE)


# and convert to dichotomy tree
new.tree2 <- ape::multi2di(new.tree1)
phy_tree(ps) <- new.tree2
phy_tree(ps)

```


Check overall how the phyla are distributed among samples. Phyloseq makes this easy


```{r}
# First aglomerate the ASVs at the phylum level using the phyloseq function, tax_glom
phylumGlommed = tax_glom(ps, "Phylum")

# There are many phyla here, so have to make a custom color palette by interpolating from an existing one in RColorBrewer
colourCount = length(table(tax_table(ps)[, "Phylum"], exclude = NULL))
getPalette = colorRampPalette(brewer.pal(9, "Spectral"))
PhylaPalette = getPalette(colourCount)

# and plot
plot_bar(phylumGlommed, x = "Sample", fill = "Phylum") + 
  scale_fill_manual(values = PhylaPalette)
```

Plot compositional (relative abundances) instead of absolute abundance using microbiome::transform

```{r}
ps_ra <- microbiome::transform(ps, transform = "compositional")
head(otu_table(ps_ra))
```

```{r}
# Then aglomerate the ASVs at the phylum level using the phyloseq function, tax_glom
phylumGlommed_RA = tax_glom(ps_ra, "Phylum")
# and plot
plot_bar(phylumGlommed_RA, x = "Sample", fill = "Phylum") + 
  scale_fill_manual(values = PhylaPalette)
```

Check abundance plots for some taxa of interest

The Proteobacteria
```{r}
# Remove the antibiotic treatments
ps_filtered <- subset_samples(ps_ra, !Treatment %in% c("Pen-Strep")) #reduced from 15 to 9 samples

# subset to proteobacteria
ps_proteo_ra <- subset_taxa(ps_filtered, Phylum == "Proteobacteria")

# define colors
colourCount = length(table(tax_table(ps_proteo_ra)[, "Order"], exclude = NULL))
getPalette = colorRampPalette(brewer.pal(9, "Spectral"))
OrderPalette = getPalette(colourCount)

# Then aglomerate the ASVs at the phylum level using the phyloseq function, tax_glom
orderGlommed_RA = tax_glom(ps_proteo_ra, "Order")
# and plot
plot_bar(orderGlommed_RA, x = "replicate", fill = "Order") + 
  scale_fill_manual(values = OrderPalette)
```
In the above you can see the increase in Vibrionales and Acetobacterales going from T0 to 20 to 25 and the decrease in Rhizobiales and Rhodobacterales


The Actinobacteriota
```{r}
# subset to proteobacteria
ps_actino_ra <- subset_taxa(ps_filtered, Phylum == "Actinobacteria")

# define colors
colourCount = length(table(tax_table(ps_actino_ra)[, "Order"], exclude = NULL))
getPalette = colorRampPalette(brewer.pal(9, "Spectral"))
OrderPalette = getPalette(colourCount)

# Then aglomerate the ASVs at the phylum level using the phyloseq function, tax_glom
orderGlommed_RA = tax_glom(ps_actino_ra, "Order")
# and plot
plot_bar(orderGlommed_RA, x = "replicate", fill = "Order") + 
  scale_fill_manual(values = OrderPalette)
```
In the above you can see a decrease in Micotrichales with increase in temperature


The Chlamydiae
```{r}
# subset to proteobacteria
ps_chlamy_ra <- subset_taxa(ps_filtered, Phylum == "Chlamydiae")

# define colors
colourCount = length(table(tax_table(ps_chlamy_ra)[, "Family"], exclude = NULL))
getPalette = colorRampPalette(brewer.pal(9, "Spectral"))
FamilyPalette = getPalette(colourCount)

# Then aglomerate the ASVs at the phylum level using the phyloseq function, tax_glom
familyGlommed_RA = tax_glom(ps_chlamy_ra, "Family")
# and plot
plot_bar(familyGlommed_RA, x = "replicate", fill = "Family") + 
  scale_fill_manual(values = FamilyPalette)
```


### Ordinations
 
#### First transform

Instead of relative abundances, need different transformation for stats. Microbiome package has both Hellinger and CLR built in. Try hellinger

```{r}
ps_hellinger <- microbiome::transform(ps, transform = "hellinger")
head(otu_table(ps_hellinger))
```


**Principal coordinate analysis (PCoA) using Bray-Curtis**
```{r}
out.pcoa <- ordinate(ps_hellinger, method = "PCoA", distance = "bray")

pcoa_plot = plot_ordination(ps_hellinger, out.pcoa, color ="Temperature", shape = "Treatment") +
  geom_point(size = 3) 
pcoa_plot
```

Scale axes by Eigenvalues

```{r}
evals <- out.pcoa$values$Eigenvalues

pcoa_plot.scaled = plot_ordination(ps_hellinger, out.pcoa, color ="Temperature", shape = "Treatment") +
  geom_point(size = 3) +
  coord_fixed(sqrt(evals[2] / evals[1]))

pcoa_plot.scaled
```


**Principal coordinate analysis (PCoA) using weighted UniFrac PCoA**
Also scaling axes by eigenvalues

```{r}
out.pcoa <- ordinate(ps_hellinger, method = "PCoA", distance = "wunifrac")

wuf_pcoa_plot = plot_ordination(ps_hellinger, out.pcoa, color ="Temperature", shape = "Treatment") +
  geom_point(size = 3) +
  coord_fixed(sqrt(evals[2] / evals[1]))


wuf_pcoa_plot
```

**Non-Metric dimensional scaling (NMDS) ordination using Bray-Curtis**

```{r}
out.nmds <- ordinate(ps_hellinger, method = "NMDS", distance = "bray")

nmds_plot = plot_ordination(ps_hellinger, out.nmds, color ="Temperature", shape = "Treatment") +
  geom_point(size = 3) 

nmds_plot
```


### Test for affects of temperature
The above shows an apparent impact of temperature. I want to find out which organisms account for this.

Filter out the antibiotic-treated samples. Those are not part of my hypothesis and would influence the interpretation of any results (as seen above).
```{r}
ps_filtered <- subset_samples(ps, !Treatment %in% c("Pen-Strep")) #reduced from 15 to 9 samples
```


#### Differential abundance with DeSeq2

This uses untransformed data, as it does its own log 2 transformation in the DeSeq calculation
```{r}
ps_deseq <- phyloseq_to_deseq2(ps_filtered, ~Temperature)
ps_deseq <- DESeq(ps_deseq)
```




*First Compare Control (Time zero) to Heat Stress (20C)*
Pull out results
```{r}
 # pulling out our results table, we specify the object, the p-value, and what contrast we want to consider by first naming the column, then the two groups we care about
deseq_res_temp_T0_20 <- results(ps_deseq, alpha=0.01, contrast=c("Temperature", "Time-zero", "20"))

# Get a glimpse at what this table currently holds with the summary command
summary(deseq_res_temp_T0_20) 
```

The above shows out of 4203 ASVs, there is 1 that increased when comparing t0 to 20 and 3 that decreased. Ie. the one that 'increased' is in higher abundance in "t0" than in "20" and the 3 that decreased are in lower abundance in t0 than in 20.


Make a table of the significant results
```{r}
# subset the table to only include those ASVs that pass the specified significance level
sigtab_deseq_res_temp_T0_20 <- deseq_res_temp_T0_20[which(deseq_res_temp_T0_20$padj < 0.01), ]

# pull out their taxonomic affiliation
sigtab_deseq_res_temp_T0_20_with_tax <- cbind(as(sigtab_deseq_res_temp_T0_20, "data.frame"), as(tax_table(ps_filtered)[row.names(sigtab_deseq_res_temp_T0_20), ], "matrix"))

# and sort that table by the baseMean column (incase there was more than one taxon)
sigtab_deseq_res_temp_T0_20_with_tax_ordered <- sigtab_deseq_res_temp_T0_20_with_tax[order(sigtab_deseq_res_temp_T0_20_with_tax$baseMean, decreasing=T), ]
sigtab_deseq_res_temp_T0_20_with_tax_ordered
```
Translation: The above table shows that the first 3 ASVs (Family: Vibrionaceae, Pseudoalteromonadaceae, and Rhodobacteraceae) which have negative log2FoldChange are in lower abundance in the T0 samples than in the 20C sample and the last ASV (Family: Thioalkalispiraceae) are in higher abundance in the T0 than 20C sample.


Check the relative abundance of of these significant ASVs to check that this translationmakes sense. 

```{r}
# pull out the ID of the ASVs that are significant
sig_taxa_T0_20_1 <- row.names(sigtab_deseq_res_temp_T0_20_with_tax_ordered)[1]

# find them in the relative abundance matrix
ind <- which(row.names(as(otu_table(ps_ra), "matrix")) == sig_taxa_T0_20_1)
as(otu_table(ps_ra), "matrix")[ind,]
```

According to this table, ASV f6e60561b232cd176707f8fb39b7cf13 (from the Vibrionaceae family) is present in samples SRR6374537, SRR6374538, SRR6374539, and SRR6374541. Looking back at the SraRunTable, these are samples 25C1, 25C2, 25C3, and 20C2. So this ASV is never present in the T0 samples... translation makes sense!


*Next Compare Heat Stress (20C) to more extreme heat stress (25C)*

Pull out results
```{r}
 # pulling out our results table, we specify the object, the p-value, and what contrast we want to consider by first naming the column, then the two groups we care about
deseq_res_temp_20_25 <- results(ps_deseq, alpha=0.01, contrast=c("Temperature", "20", "25"))

# Get a glimpse at what this table currently holds with the summary command
summary(deseq_res_temp_20_25) 
```

Translation: out of 1451 ASVs, there is 1 that increased when comparing temperature "20" to "25" and 1 that decreased. Ie. the one that 'increased' is in higher abundance in "20" than in "25"  and the one that decreased is in lower abundance in "20" than "25."


Make a table of the significant results
```{r}
# subset the table to only include those ASVs that pass the specified significance level
sigtab_deseq_res_temp_20_25 <- deseq_res_temp_20_25[which(deseq_res_temp_20_25$padj < 0.01), ]

# pull out their taxonomic affiliation
sigtab_deseq_res_temp_20_25_with_tax <- cbind(as(sigtab_deseq_res_temp_20_25, "data.frame"), as(tax_table(ps_filtered)[row.names(sigtab_deseq_res_temp_20_25), ], "matrix"))

# and sort that table by the baseMean column (incase there was more than one taxon)
sigtab_deseq_res_temp_20_25_with_tax[order(sigtab_deseq_res_temp_20_25_with_tax$baseMean, decreasing=T), ]

```
The ASV that is in lower abundance in 20 than 25 is (negative log2FoldChange) is Acidobacter>...>Acanthopleuribacter and the one that is in higher abundance in 20 than 25 is Bacteroidetes>...>Aquimarina.




*Next Compare Control (Time zero) to extreme heat stress (25C)*

Pull out results
```{r}
 # pulling out our results table, we specify the object, the p-value, and what contrast we want to consider by first naming the column, then the two groups we care about
deseq_res_temp_T0_25 <- results(ps_deseq, alpha=0.01, contrast=c("Temperature", "Time-zero", "25"))

# Get a glimpse at what this table currently holds with the summary command
summary(deseq_res_temp_T0_25) 
```

Translation: out of 4203 ASVs, there are2 that increased when comparing temperature "T0" to "25" and 11 that decreased. Ie. the one that 'increased' is in higher abundance in "T0" than in "25" and those that decreased are in lower abundance in T0 than in 25.


Make a table of the significant results
```{r}
# subset the table to only include those ASVs that pass the specified significance level
sigtab_deseq_res_temp_T0_25 <- deseq_res_temp_T0_25[which(deseq_res_temp_T0_25$padj < 0.01), ]

# pull out their taxonomic affiliation
sigtab_deseq_res_temp_T0_25_with_tax <- cbind(as(sigtab_deseq_res_temp_T0_25, "data.frame"), as(tax_table(ps_filtered)[row.names(sigtab_deseq_res_temp_T0_25), ], "matrix"))

# and sort that table by the baseMean column (incase there was more than one taxon)
sigtab_deseq_res_temp_T0_25_with_tax[order(sigtab_deseq_res_temp_T0_25_with_tax$baseMean, decreasing=T), ]

```
There are lots of Vibrionaceae here that are higher in the heat stressed (25C) samples than T0. Also Pseudoalteromonadaceae, Nitrincolaceae, Rhodobacteraceae, 


Since there are several ASVs here, you can also visualize these results in a plot
```{r}
# Incase you have many ASVs, define enough colors for the plot
theme_set(theme_bw())
scale_fill_discrete <- function(palname = "Set1", ...) {
  scale_fill_brewer(palette = palname, ...)
}

# And plot the families, color coding by Class
x = tapply(sigtab_deseq_res_temp_T0_25_with_tax$log2FoldChange, sigtab_deseq_res_temp_T0_25_with_tax$Family, function(x) max(x))
x = sort(x, TRUE)
sigtab_deseq_res_temp_T0_25_with_tax$Family = factor(as.character(sigtab_deseq_res_temp_T0_25_with_tax$Family), levels=names(x))
ggplot(sigtab_deseq_res_temp_T0_25_with_tax, aes(x=Family, y=log2FoldChange, color=Class)) + geom_point(size=6) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
```

#### Export
Export the data tables as count tables (with no transformation) and without any aggregation.

For this study, I am excluding the antiobiotic samples, as those don't really fir the meta-analysis study. I only want the samples from different  temperatures (time zero, 20, and 25C).

First, filter out any ASVs without a species or genus and agglomerate at the genus-level. This will allow for comparisons to other datasets

```{r}
# Make a brief descriptive file
descript_file<-writeLines(c("The phyloseq object in this file are data processed from Green et al. 2019 (doi: 10.1007/s00248-018-1242-9), NCBI BioProject # PRJNA421986.
The ASV count table was produced using Qiime2 (v. 2020.02), calling DADA2 for denoising, merging, and ASV inference, and a Silva v132 Naive 
Bayes classifier for taxonomy calling.
Samples were collected from Pacific oyster (Crassostrea gigas) samples in Australia. AUthors performed a simulated heat wave in mesocosms to test
impacts on microbiome.
The phyloseq object, ps_filtered, was filtered to remove a subset of samples that were also used to test the impact of anitibiotics. The remaining 
samples here were not treated with antibiotics. The remaining samples correspond to a time-zero (field) collection and a time-4day collection from two 
different heat stress conditions (20ËšC and 25ËšC). Each assay has biological triplicates and thus there are 9 total samples.
On day 4 in the 25ËšC treatment, the oysters were showing signs of mortality.
The metadata in this file come from the SraRunTable and have been modified to clearly indicate temperature and mortality data.
The otu_table object in the phyloseq object contains absolute abundaces of ASVs, without any tranformation or agglomeration at higher taxonomic levels.

qiime2 analysis file (done in Jupyter notebook) =  Qiime2_PRJNA421986.ipynb

processing of qiime2 and preliminary stats file (done in R) = PRJNA421986_PacificOyster_QIIME2_postanalysis.Rmd"))

# Export as .RData
save(ps_filtered, descript_file, file = "../../Processed/PRJNA421986_qiime2_processed_data.RData")

```


